File: useless/p1_java/synteza/zad1/Phonemes.java
package synteza.zad1;

import java.util.regex.Pattern;
import java.util.regex.Matcher;
import java.util.Arrays;
import java.util.List;
import java.util.ArrayList;
import java.util.Map;
import java.util.HashMap;
import java.util.Collections;

public class Phonemes {
	private static final List<String> Palatalizables;
	static {
		String[] a = {"c","dz","n","s","z"};
		Palatalizables = Arrays.asList(a);
	}
	private static final List<String> Digraphs;
	static {
		String[] a = {"ch","cz","drz","dz","dzi","rz","sz"};
		Digraphs = Arrays.asList(a);
	}
	private static final List<String> Vowels;
	static {
		String[] a = {"a","on","e","en","i","o","u","y"};
		Vowels = Arrays.asList(a);
	}
	private static final Map<String, String> PolishLetters;
	static {
		Map<String, String> map = new HashMap<String, String>();
		map.put("ą", "on");
		map.put("ć", "ci");
		map.put("ę", "en");
		map.put("ł", "ll");
		map.put("ń", "ni");
		map.put("ó", "u");
		map.put("ś", "si");
		map.put("ź", "zi");
		map.put("ż", "rz");
		PolishLetters = Collections.unmodifiableMap(map);
	}
	private static final Map<String, String> VoicelessEquivalents;
	static {
		Map<String, String> map = new HashMap<String, String>();
		map.put("b", "p");
		map.put("d", "t");
		map.put("w", "f");
		map.put("g", "k");
		map.put("z", "s");
		map.put("rz", "sz");
		map.put("drz", "cz");
		map.put("zi", "si");
		map.put("dz", "c");
		map.put("dzi", "ci");
		VoicelessEquivalents = Collections.unmodifiableMap(map);
	}

	private static final Map<String, String> VoicedEquivalents;
	static {
		Map<String, String> map = new HashMap<String, String>();
		for (String key : VoicelessEquivalents.keySet()) {
			map.put(VoicelessEquivalents.get(key), key);
		}
		VoicedEquivalents = Collections.unmodifiableMap(map);
	}

	public static ArrayList<String> lineToPhonemes(String line) {
		line = line.toLowerCase();
		ArrayList<String> phonemes = new ArrayList<String>();
		String[] words = line.split(" ");
		for (String word : words) {
			phonemes.addAll(wordToPhonemes(word));
			phonemes.add(" ");
		}
		return phonemes;
	}

	public static ArrayList<String> wordToPhonemes(String word) {
		word = word.toLowerCase();
		ArrayList<String> phonemes = new ArrayList<String>();
		if(word.length()>0) {
			for(int i=0; i<word.length(); i++) {
				String letter = word.substring(i, i+1);
				String phoneme;
				if((phoneme = PolishLetters.get(letter)) == null) {
					phoneme = letter;
				}

				phonemes.add(phoneme);
			}

			connectDigraphs(phonemes);
			palatalize(phonemes);
			lengthenI(phonemes);
			unnasalizeVowels(phonemes);
			splitNosalVowels(phonemes);
			fixLast(phonemes);
			backwardVoiceless(phonemes);
			forwardVoiceless(phonemes);
			backwardVoiced(phonemes);
			fixLetterU(phonemes);
		}
		return phonemes;
	}

	private static void connectDigraphs(ArrayList<String> phonemes) {
		for(int i=phonemes.size()-1; i>0; i--) {
			if(isDigraph(phonemes.get(i-1), phonemes.get(i))) {
				phonemes.set(i-1, phonemes.get(i-1)+phonemes.get(i));
				phonemes.remove(i);
				i--;
			}
		}
	}

	private static void palatalize(ArrayList<String> phonemes) {
		for(int i=phonemes.size()-1; i>0; i--) {
			if(phonemes.get(i).equals("i") && isPalatalizable(phonemes.get(i-1))) {
				phonemes.set(i-1, phonemes.get(i-1)+"i");
				if(i+1<phonemes.size() && isVowel(phonemes.get(i+1))) {
					phonemes.remove(i);
				}
				i--;
			}
		}
	}

	private static final List<String> prefixesWithAu;
	static {
		String[] a = {"nau","zau","prau","pozau","nienau","niezau","nieprau","niepozau"};
		prefixesWithAu = Arrays.asList(a);
	}

	private static boolean fixAuPrefix(String word) {
		for(String prefix : prefixesWithAu) {
			if(word.startsWith(prefix)) {
				// exceptions:
				//	.*au.o.*
				//	.*au.$
				if(word.length()==prefix.length()+1 || (word.length()>prefix.length()+1 && word.substring(prefix.length()+1, prefix.length()+2).equals("o"))) {
					return false;
				}
				return true;
			}
		}
		return false;
	}

	private static final List<String> prefixesWithEu;
	static {
		String[] a = {"przeu","nieprzeu"};
		prefixesWithEu = Arrays.asList(a);
	}

	private static boolean fixEuPrefix(String word) {
		for(String prefix : prefixesWithEu) {
			if(word.startsWith(prefix)) {
				return true;
			}
		}
		return false;
	}

	// exceptions:
	//	.*eusz.*
	//	.*eusi.*
	//	.*euj.*
	//	.*eu.$
	//	.*aurk.*
	private static final List<String> euExceptionalSuccessors;
	static {
		String[] a = {"sz","si","j"};
		euExceptionalSuccessors = Arrays.asList(a);
	}
	private static void fixLetterU(ArrayList<String> phonemes) {
		String word = concat(phonemes);
		boolean skipFirst = fixAuPrefix(word) || fixEuPrefix(word);
		for(int i=1; i<phonemes.size(); i++) {
			if(phonemes.get(i).equals("u") && (phonemes.get(i-1).equals("a")||phonemes.get(i-1).equals("e"))) {
				if(skipFirst) {
					skipFirst = false;
					continue;
				}
				if(phonemes.get(i-1).equals("e")) {
					if(i+2==phonemes.size()) break;
					if(i+1<phonemes.size() && euExceptionalSuccessors.contains(phonemes.get(i+1))) continue;
				} else {
					if(i+2<phonemes.size() && (phonemes.get(i+1)+phonemes.get(i+2)).equals("rk")) continue;
				}
				phonemes.set(i, "ll");
			}
		}
	}


	private static void lengthenI(ArrayList<String> phonemes) {
		for(int i=0; i<phonemes.size()-1; i++) {
			if(phonemes.get(i).equals("i") && isVowel(phonemes.get(i+1))) {
				phonemes.set(i, "j");
				i++;
			}
		}
	}

	private static void unnasalizeVowels(ArrayList<String> phonemes) {
		for(int i=0; i<phonemes.size()-1; i++) {
			String voiceless;
			if((phonemes.get(i).equals("on") || phonemes.get(i).equals("en")) && (phonemes.get(i+1).equals("l") || phonemes.get(i+1).equals("ll"))) {
				phonemes.set(i, phonemes.get(i).substring(0,1));
				i++;
			}
		}
	}

	// TODO: dz, c (??)
	private static final List<String> nosalVowelsSplitters;
	static {
		String[] a = {"t","d","b","p","ci","dzi","dz","c"};
		nosalVowelsSplitters = Arrays.asList(a);
	}
	private static void splitNosalVowels(ArrayList<String> phonemes) {
		for(int i=0; i<phonemes.size()-1; i++) {
			String voiceless;
			if((phonemes.get(i).equals("on") || phonemes.get(i).equals("en")) && nosalVowelsSplitters.contains(phonemes.get(i+1))) {
				if(phonemes.get(i+1).equals("b")) {
					phonemes.add(i+1, "m");
				} else {
					phonemes.add(i+1, phonemes.get(i).substring(1,2));					
				}
				phonemes.set(i, phonemes.get(i).substring(0,1));
				i+=2;
			}
		}
	}

	private static void fixLast(ArrayList<String> phonemes) {
		String voiceless;
		if(phonemes.size()>1) {
			if((voiceless=VoicelessEquivalents.get(phonemes.get(phonemes.size()-1)))!=null) {
				phonemes.set(phonemes.size()-1, voiceless);
			}
			if(phonemes.get(phonemes.size()-1).equals("en")) {
				phonemes.set(phonemes.size()-1, "e");
			}
		}
	}

	private static void backwardVoiceless(ArrayList<String> phonemes) {
		for(int i=0; i<phonemes.size()-1; i++) {
			String voiceless;
			if((voiceless=VoicelessEquivalents.get(phonemes.get(i)))!=null && VoicedEquivalents.get(phonemes.get(i+1))!=null) {
				phonemes.set(i, voiceless);
				i++;
			}
		}
	}

	private static void forwardVoiceless(ArrayList<String> phonemes) {
		for(int i=phonemes.size()-1; i>0; i--) {
			String voiceless;
			if(VoicedEquivalents.get(phonemes.get(i-1))!=null && (voiceless=VoicelessEquivalents.get(phonemes.get(i)))!=null && (voiceless.equals("f")||voiceless.equals("sz"))) {
				phonemes.set(i, voiceless);
				i--;
			}
		}
	}

	private static void backwardVoiced(ArrayList<String> phonemes) {
		for(int i=0; i<phonemes.size()-1; i++) {
			String voiceless;
			if((voiceless=VoicedEquivalents.get(phonemes.get(i)))!=null && VoicelessEquivalents.get(phonemes.get(i+1))!=null) {
				phonemes.set(i, voiceless);
				i++;
			}
		}
	}

	public static boolean isDigraph(String l1, String l2) {
		String d=l1+l2;
		return Digraphs.contains(d);
	}

	public static boolean isPalatalizable(String l) {
		return Palatalizables.contains(l);
	}

	public static boolean isVowel(String l) {
		return Vowels.contains(l);
	}

	public static boolean isConsonant(String l) {
		return !isVowel(l);
	}

	private static String concat(ArrayList<String> al) {
		String str = "";
		for(String s : al) {
			str+=s;
		}
		return str;
	}

	public static boolean isPhoneme(String l) {
		return Pattern.compile("[a-z]{1,3}").matcher(l).find();
	}
}

File: p1/Phonemes.py
# -*- coding: utf-8 -*-

import re

class Phonemes(object):
	PALATALIZABLES = ["c","dz","n","s","z"]
	DIGRAPHS = ["ch","cz","drz","dz","dzi","rz","sz"]
	VOWELS = ["a","on","e","en","i","o","u","y"]
	POLISH_LETTERS = {
		u"ą":	"on",
		u"ć":	"ci",
		u"ę":	"en",
		u"ł":	"ll",
		u"ń":	"ni",
		u"ó":	"u",
		u"ś":	"si",
		u"ź":	"zi",
		u"ż":	"rz"}
	VOICELESS_EQUIVALENTS = {
		"b":	"p",
		"d":	"t",
		"w":	"f",
		"g":	"k",
		"z":	"s",
		"rz":	"sz",
		"drz":	"cz",
		"zi":	"si",
		"dz":	"c",
		"dzi":	"ci"}
	VOICED_EQUIVALENTS = {
		"p":	"b",
		"t":	"d",
		"f":	"w",
		"k":	"g",
		"s":	"z",
		"sz":	"rz",
		"cz":	"drz",
		"si":	"zi",
		"c":	"dz",
		"ci":	"dzi"}
	@classmethod
	def line_to_phonemes(cls, line):
		line = line.lower()
		phonemes = []
		words = line.split()
		for word in words:
			phonemes += cls.word_to_phonemes(word)
			phonemes += [" "]

		if(len(phonemes)>1):
			del phonemes[len(phonemes)-1]
		return phonemes

	@classmethod
	def word_to_phonemes(cls, word):
		word = word.lower()
		phonemes = []
		if len(word) > 0:
			for i in range(0,len(word)):
				letter = word[i:i+1]
				if cls.POLISH_LETTERS.has_key(letter):
					phonemes += [cls.POLISH_LETTERS[letter]]
				else:
					if(letter == u"x"):
						phonemes += ["k", "s"]
					else:
						phonemes += [letter.encode("ascii","ignore")]
			cls.connect_digraphs(phonemes);
			cls.palatalize(phonemes);
			cls.lengthen_i(phonemes);
			cls.unnasalize_vowels(phonemes);
			cls.split_nosal_vowels(phonemes);
			cls.fix_last(phonemes);
			cls.backward_voiceless(phonemes);
			cls.forward_voiceless(phonemes);
			cls.backward_voiced(phonemes);
			cls.fix_letter_u(phonemes);
		return phonemes

	@classmethod
	def connect_digraphs(cls, phonemes):
		i = len(phonemes)-1
		while i > 0:
			if cls.is_digraph(phonemes[i-1], phonemes[i]):
				phonemes[i-1] = phonemes[i-1]+phonemes[i]
				if(phonemes[i-1] == "ch"):
					phonemes[i-1] = "h"
				del phonemes[i]
			i-=1

	@classmethod
	def palatalize(cls, phonemes):
		i = len(phonemes)-1
		while i > 0:
			if phonemes[i]=="i" and cls.is_palatalizable(phonemes[i-1]):
				phonemes[i-1] = phonemes[i-1]+"i"
				if i+1 < len(phonemes) and cls.is_vowel(phonemes[i+1]):
					del phonemes[i]
				i-=1
			i-=1

	@classmethod
	def lengthen_i(cls, phonemes):
		i = 0
		while i<len(phonemes)-1:
			if phonemes[i]=="i" and cls.is_vowel(phonemes[i+1]):
				phonemes[i] = "j"
				i+=1;
			i+=1

	@classmethod
	def unnasalize_vowels(cls, phonemes):
		i = 0
		while i<len(phonemes)-1:
			if phonemes[i] in ["on", "en"] and phonemes[i+1] in ["l", "ll"]:
				phonemes[i] = phonemes[i][0:1]
				i+=1
			i+=1

	@classmethod
	def fix_last(cls, phonemes):
		if len(phonemes)>1:
			if cls.VOICELESS_EQUIVALENTS.has_key(phonemes[-1]):
				phonemes[-1] = cls.VOICELESS_EQUIVALENTS[phonemes[-1]]
			if phonemes[-1] == "en":
				phonemes[-1] = "e"
		if len(phonemes)>2:
			if phonemes[-1] == "i" and (phonemes[-2] in ["i", "j"]):
				del phonemes[-1]
				phonemes[-1] = "i"


	@classmethod
	def backward_voiceless(cls, phonemes):
		i = 0
		while i < len(phonemes)-1:
			if cls.VOICELESS_EQUIVALENTS.has_key(phonemes[i]) and cls.VOICED_EQUIVALENTS.has_key(phonemes[i+1]):
				phonemes[i] = cls.VOICELESS_EQUIVALENTS[phonemes[i]]
				i+=1
			i+=1

	@classmethod
	def forward_voiceless(cls, phonemes):
		i = len(phonemes)-1
		while i > 0:
			if cls.VOICED_EQUIVALENTS.has_key(phonemes[i-1]) and phonemes[i] in ["w", "rz"]:
				phonemes[i] = cls.VOICELESS_EQUIVALENTS[phonemes[i]]
				i-=1
			i-=1

	@classmethod
	def backward_voiced(cls, phonemes):
		i = 0
		while i < len(phonemes)-1:
			if cls.VOICED_EQUIVALENTS.has_key(phonemes[i]) and cls.VOICELESS_EQUIVALENTS.has_key(phonemes[i+1]):
				phonemes[i] = cls.VOICED_EQUIVALENTS[phonemes[i]]
				i+=1
			i+=1

	NOSAL_VOWEL_SPLITTERS = ["t","d","b","p","ci","dzi","dz","c"]
	@classmethod
	def split_nosal_vowels(cls, phonemes):
		i = 0
		while i<len(phonemes)-1:
			if phonemes[i] in ["on","en"] and phonemes[i+1] in cls.NOSAL_VOWEL_SPLITTERS:
				if phonemes[i+1] in ["b", "p"]:
					phonemes.insert(i+1, "m");
				else:
					if cls.is_palatalized(phonemes[i+1]):
						phonemes.insert(i+1, "ni")
					else:
						phonemes.insert(i+1, "n")


				phonemes[i] = phonemes[i][0:1]
				i+=2
			i+=1

	PREFIXES_WITH_AU = ["nau","zau","prau", "unau", "pozau","nienau","niezau","nieprau", "nieunau", "niepozau"]
	@classmethod
	def has_au_prefix_exception(cls, word):
		for prefix in cls.PREFIXES_WITH_AU:
			if re.match(prefix, word):
				# exceptions:
				#	.*au.o.*
				#	.*au.$
				return not (len(word)==len(prefix)+1) # or (len(word)>len(prefix)+1 and word[len(prefix)+1:len(prefix)+2]=="o"))
		return False

	PREFIXES_WITH_EU = ["przeu","nieprzeu"]
	@classmethod
	def has_eu_prefix_exception(cls, word):
		for prefix in cls.PREFIXES_WITH_EU:
			if re.match(prefix, word):
				return True
		return False

	# exceptions:
	#	.*eusz.*
	#	.*eusi.*
	#	.*euj.*
	#	.*eu.$
	#	.*aurk.*
	EU_EXCEPTIONAL_SUCCESSORS = ["sz","si","j"]
	@classmethod
	def fix_letter_u(cls, phonemes):
		word = "".join(phonemes)
		skip_first = cls.has_au_prefix_exception(word) or cls.has_eu_prefix_exception(word);
		for i in range(1,len(phonemes)):
			if phonemes[i]=="u" and phonemes[i-1] in ["a","e"]:
				if skip_first:
					skip_first = False
					continue
				if phonemes[i-1]=="e":
					if i+2 == len(phonemes):
						break
					if i+1 < len(phonemes) and phonemes[i+1] in cls.EU_EXCEPTIONAL_SUCCESSORS:
						continue

				else:
					if i+2 < len(phonemes) and phonemes[i+1]+phonemes[i+2]=="rk":
						continue
				phonemes[i] = "ll"

	@classmethod
	def is_digraph(cls, l1, l2):
		d = l1+l2
		return d in cls.DIGRAPHS

	@classmethod
	def is_palatalizable(cls, l):
		return l in cls.PALATALIZABLES

	@classmethod
	def is_palatalized(cls, l):
		return re.match("[a-z]+i$", l)

	@classmethod
	def is_vowel(cls, l):
		return l in cls.VOWELS

	@classmethod
	def is_consonant(cls, l):
		return not cls.is_vowel(l)

	@classmethod
	def is_phoneme(cls, l):
		return re.match("[a-z]{1,3}$", l)


